jl4-core/src/L4/TypeCheck/Types.hs

```diff
@@ -210,3 +212,114 @@ instance Alternative Check where
 
   (<|>) :: Check a -> Check a -> Check a
   (<|>) m1 m2 = MkCheck $ \e s -> runCheck m1 e s ++ runCheck m2 e s
+
+-- ----------------------------------------------------------------------------
+-- Primitives for Uniques, Errors and similar.
+-- ----------------------------------------------------------------------------
+
+step :: Check Int
+step = do
+  current <- use #supply
+  let next = current + 1
+  assign #supply next
+  pure current
+
+newUnique :: Check Unique
+newUnique = do
+  i <- step
+  u <- asks (.moduleUri)
+  pure (MkUnique 'c' i u)
+
+addError :: CheckError -> Check ()
+addError e = do
+  ctx <- use #errorContext
+  with (MkCheckErrorWithContext e ctx)
+
+choose :: [Check a] -> Check a
+choose = asum
+
+anyOf :: [a] -> Check a
+anyOf = asum . fmap pure
```

Comment:
this definition is bogus, (or at least misleadingly named and defined) I think.

Reply:
it takes the first definition that succeeds, which must be the first element in the list because `pure x` always succeeds. I think it's equivalent to 
```haskell
\case [] -> empty; (x : _) -> pure x
```

Reply:
@kosmikus ^